File: web-change-scanner/code/storage.js
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';

const NATS_ENABLED = process.env.NATS_ENABLED;
const OUT_DIR = process.env.OUT_DIR || '/out';
const NATS_URL = process.env.NATS_URL || 'nats://localhost:4222';
const BUCKET = 'web-change-scanner';

function sanitizeDomain(domain) {
  return domain.replace(/\./g, '_').replace(/[^a-zA-Z0-9_-]/g, '');
}

function filePath(domain) {
  return path.join(OUT_DIR, 'web-change-scanner.' + domain + '.baseline.json');
}

async function fsLoad(domain) {
  const fp = filePath(domain);
  try {
    const raw = await readFile(fp, 'utf8');
    return JSON.parse(raw);
  } catch (err) {
    if (err.code === 'ENOENT') return null;
    process.stderr.write(`[storage] warning: failed to read ${fp}: ${err.message}\n`);
    return null;
  }
}

async function fsSave(domain, snapshot) {
  await mkdir(OUT_DIR, { recursive: true });
  const fp = filePath(domain);
  await writeFile(fp, JSON.stringify(snapshot, null, 2));
}

async function natsLoad(domain) {
  const { connect, StringCodec } = await import('nats');
  const sc = StringCodec();
  const key = sanitizeDomain(domain);
  const nc = await connect({ servers: NATS_URL });
  try {
    const js = nc.jetstream();
    const kv = await js.views.kv(BUCKET, { history: 1 });
    const entry = await kv.get(key);
    const value = entry ? sc.decode(entry.value) : null;
    return value ? JSON.parse(value) : null;
  } finally {
    await nc.drain();
  }
}

async function natsSave(domain, snapshot) {
  const { connect, StringCodec } = await import('nats');
  const sc = StringCodec();
  const key = sanitizeDomain(domain);
  const nc = await connect({ servers: NATS_URL });
  try {
    const js = nc.jetstream();
    const kv = await js.views.kv(BUCKET, { history: 1 });
    await kv.put(key, sc.encode(JSON.stringify(snapshot)));
  } finally {
    await nc.drain();
  }
}

export async function load(domain) {
  if (NATS_ENABLED !== 'False') {
    try {
      return await natsLoad(domain);
    } catch (err) {
      process.stderr.write(`[storage] warning: NATS load failed, falling back to filesystem: ${err.message}\n`);
      return await fsLoad(domain);
    }
  }
  return await fsLoad(domain);
}

export async function save(domain, snapshot) {
  if (NATS_ENABLED !== 'False') {
    try {
      return await natsSave(domain, snapshot);
    } catch (err) {
      process.stderr.write(`[storage] warning: NATS save failed, falling back to filesystem: ${err.message}\n`);
      return await fsSave(domain, snapshot);
    }
  }
  return await fsSave(domain, snapshot);
}

File: network-change-scanner/code/storage.py
import os
import re
import sys
import json
import asyncio

NATS_ENABLED = os.getenv('NATS_ENABLED', 'False')
OUT_DIR = os.getenv('OUT_DIR', '/out')
NATS_URL = os.getenv('NATS_URL', 'nats://localhost:4222')
BUCKET = 'network-change-scanner'


def sanitize_key(key: str) -> str:
    key = re.sub(r'[.:]', '_', key)
    return re.sub(r'[^a-zA-Z0-9_-]', '', key)


def _file_path(target_key: str) -> str:
    return os.path.join(OUT_DIR, f'network-change-scanner.{target_key}.baseline.json')


# --- Filesystem mode ---

def _fs_load(target_key: str) -> dict | None:
    fp = _file_path(target_key)
    try:
        with open(fp, 'r', encoding='utf-8') as fh:
            return json.loads(fh.read())
    except FileNotFoundError:
        return None
    except Exception as e:
        sys.stderr.write(f'[storage] warning: failed to read {fp}: {e}\n')
        return None


def _fs_save(target_key: str, snapshot: dict) -> None:
    os.makedirs(OUT_DIR, exist_ok=True)
    fp = _file_path(target_key)
    with open(fp, 'w', encoding='utf-8') as fh:
        fh.write(json.dumps(snapshot, indent=2))


# --- NATS KV mode ---

async def _nats_load(target_key: str) -> dict | None:
    import nats
    import nats.js.api

    key = sanitize_key(target_key)
    nc = await nats.connect(NATS_URL)
    try:
        js = nc.jetstream()
        try:
            kv = await js.key_value(BUCKET)
        except Exception:
            kv = await js.create_key_value(
                nats.js.api.KeyValueConfig(bucket=BUCKET, history=1)
            )
        try:
            entry = await kv.get(key)
            return json.loads(entry.value.decode())
        except Exception:
            return None
    finally:
        await nc.drain()


async def _nats_save(target_key: str, snapshot: dict) -> None:
    import nats
    import nats.js.api

    key = sanitize_key(target_key)
    nc = await nats.connect(NATS_URL)
    try:
        js = nc.jetstream()
        try:
            kv = await js.key_value(BUCKET)
        except Exception:
            kv = await js.create_key_value(
                nats.js.api.KeyValueConfig(bucket=BUCKET, history=1)
            )
        await kv.put(key, json.dumps(snapshot).encode())
    finally:
        await nc.drain()


# --- Public interface ---

def load(target_key: str) -> dict | None:
    if NATS_ENABLED != 'False':
        try:
            return asyncio.run(_nats_load(target_key))
        except Exception as e:
            sys.stderr.write(
                f'[storage] warning: NATS load failed, falling back to filesystem: {e}\n'
            )
            return _fs_load(target_key)
    return _fs_load(target_key)


def save(target_key: str, snapshot: dict) -> None:
    if NATS_ENABLED != 'False':
        try:
            asyncio.run(_nats_save(target_key, snapshot))
            return
        except Exception as e:
            sys.stderr.write(
                f'[storage] warning: NATS save failed, falling back to filesystem: {e}\n'
            )
            _fs_save(target_key, snapshot)
            return
    _fs_save(target_key, snapshot)
